## Analyze Data and Prepare for Modeling 

Dataset contains customer attributes like age, occupation, marital status, education, communication channel, call timing/duration, previous campaign outcome, and conversion status, enabling analysis of customer interactions and campaign effectiveness in insurance.

Each row represents a customer interaction record, detailing various attributes such as age, occupation, marital status, education level, communication channel, call timing and duration, previous campaign outcome, and conversion status. This dataset can be used for analyzing customer behavior, predicting conversion rates, and optimizing marketing strategies in the insurance industry.

Let's start by examining the dataset for any blank or duplicate values and then prepare it for a descriptive statistical analysis. I'll load the dataset first to see what we're working with.

In [20]:
import pandas as pd

# Load the dataset
file_path = '../data/Bank Marketing Campaign Dataset.csv'
data = pd.read_csv(file_path)

# Display the first few rows of the dataset and its basic information
data.head(), data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45211 entries, 0 to 45210
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   occupation                 45211 non-null  object
 1   age                        45211 non-null  int64 
 2   education_level            45211 non-null  object
 3   marital_status             45211 non-null  object
 4   communication_channel      45211 non-null  object
 5   call_month                 45211 non-null  object
 6   call_day                   45211 non-null  int64 
 7   call_duration              45211 non-null  int64 
 8   call_frequency             45211 non-null  int64 
 9   previous_campaign_outcome  45211 non-null  object
 10  conversion_status          45211 non-null  object
dtypes: int64(4), object(7)
memory usage: 3.8+ MB


(             occupation  age education_level marital_status  \
 0  administrative_staff   28     high_school        married   
 1  administrative_staff   58    unidentified        married   
 2               jobless   40     high_school       divorced   
 3        retired_worker   63     high_school        married   
 4        business_owner   43         college        married   
 
   communication_channel call_month  call_day  call_duration  call_frequency  \
 0          unidentified  September         9              1               1   
 1          unidentified       June         5            307               2   
 2                mobile   February         4            113               1   
 3                mobile      April         7             72               5   
 4              landline       July        29            184               4   
 
   previous_campaign_outcome conversion_status  
 0                successful     not_converted  
 1              unidentified     n

The dataset consists of 45,211 entries across 11 columns with various attributes such as occupation, age, education level, marital status, communication channel, and others related to a marketing campaign.

Here's a brief overview of the columns:

- occupation: Occupation of the customer.
- age: Age of the customer.
- education_level: Educational level of the customer.
- marital_status: Marital status of the customer.
- communication_channel: Communication channel used for the call.
- call_month: Month when the call was made.
- call_day: Day of the month when the call was made.
- call_duration: Duration of the call in minutes.
- call_frequency: Frequency of the calls made.
- previous_campaign_outcome: Outcome of the previous marketing campaign.
- conversion_status: Whether the customer converted or not after the call.

Next, I'll check for any duplicate entries and blank or missing values in the dataset. ​​

The dataset contains 6 duplicate entries and no missing values across all columns.

In [21]:
# Check for duplicate entries
duplicate_count = data.duplicated().sum()

# Check for any missing or blank values
missing_values = data.isna().sum()
print("Duplicate entries:", duplicate_count
    , "\nMissing values:\n",missing_values)

Duplicate entries: 6 
Missing values:
 occupation                   0
age                          0
education_level              0
marital_status               0
communication_channel        0
call_month                   0
call_day                     0
call_duration                0
call_frequency               0
previous_campaign_outcome    0
conversion_status            0
dtype: int64


The next steps would be:

1. Remove the duplicate entries to clean the data.
2. Perform a descriptive statistical analysis to understand the distribution and summary statistics of the numerical data.

In [22]:
# Remove duplicate entries
cleaned_data = data.drop_duplicates()

# Verify removal
remaining_duplicates = cleaned_data.duplicated().sum()
print("Duplicates value don't exist." if remaining_duplicates == 0 else f"There are {remaining_duplicates} duplicates.")


Duplicates value don't exist.


Let's proceed to the descriptive statistical analysis.

I will calculate the basic statistics for the numerical columns: age, call_day, call_duration, and call_frequency.

This includes measures like mean, median, standard deviation, and range.

In [23]:
# Descriptive statistics for numerical columns
numerical_description = cleaned_data[['age', 'call_day', 'call_duration', 'call_frequency']].describe()
numerical_description

age     call_day  call_duration  call_frequency
count  45205.000000  45205.00000   45205.000000    45205.000000
mean      40.937087     15.80688     258.183055        2.763898
std       10.619130      8.32234     257.538504        3.098189
min       18.000000      1.00000       0.000000        1.000000
25%       33.000000      8.00000     103.000000        1.000000
50%       39.000000     16.00000     180.000000        2.000000
75%       48.000000     21.00000     319.000000        3.000000
max       95.000000     31.00000    4918.000000       63.000000

#### Age:
- Mean: Approximately 40.94 years
- Standard Deviation: Approximately 10.62 years
- Min/Max: Ranges from 18 to 95 years
- Quartiles: The 25th, 50th (median), and 75th percentiles are 33, 39, and 48 years, respectively.

#### Call Day (day of the month when the call was made):
- Mean: Approximately 15.81
- Standard Deviation: Approximately 8.32
- Min/Max: Ranges from 1 to 31
- Quartiles: The 25th, 50th (median), and 75th percentiles are 8, 16, and 21, respectively.

#### Call Duration (in seconds):
- Mean: Approximately 258.18 seconds
- Standard Deviation: Approximately 257.54 seconds
- Min/Max: Ranges from 0 to 4918 seconds
- Quartiles: The 25th, 50th (median), and 75th percentiles are 103, 180, and 319 seconds, respectively.

#### Call Frequency:
- Mean: Approximately 2.76 calls
- Standard Deviation: Approximately 3.10 calls
- Min/Max: Ranges from 1 to 63 calls. Interesting! 63 calls is the highest frequency. Why? :)
- Quartiles: The 25th, 50th (median), and 75th percentiles are 1, 2, and 3 calls, respectively.

### Plan for next steps:
#### Perform analysis for understanding the effectiveness of the bank marketing campaigns and for strategic planning

In [24]:
# Calculate overall conversion rate
data = cleaned_data # Replace with your cleaned dataframe for more clean code   

conversion_rate_overall = data['conversion_status'].value_counts(normalize=True)['converted'] * 100
print("""Overall Conversion Rate: {}%""".format(round(conversion_rate_overall, 2)))

Overall Conversion Rate: 11.7%


The overall conversion rate in the dataset is approximately __11.70%__. This indicates that about 11.7% of the contacts in the campaign resulted in a conversion.

### Conversion Rate by different demographic attributes

Initial we will set up a function to calculate the conversion rate by different demographic attributes.

In [34]:
# Define conversion rate function
def conversion_rate_by(attribute):
    conversion_rates = data.groupby(attribute, observed=True)['conversion_status'].value_counts(normalize=True).unstack().fillna(0)
    conversion_rates['conversion_rate'] = conversion_rates['converted'] * 100
    return conversion_rates[['conversion_rate']]

## By Ocupation

In [35]:
conversion_rate_occupation = conversion_rate_by('occupation')
print("Conversion rates by occupation (%):")
conversion_rate_occupation.head(20).sort_values(by='conversion_rate', ascending=False)

Conversion rates by occupation (%):


conversion_status     conversion_rate
occupation                           
student                     28.678038
retired_worker              22.791519
jobless                     15.502686
executive                   13.757005
administrative_staff        12.205029
independent_worker          11.842939
unidentified                11.805556
technical_specialist        11.058452
service_worker               8.885143
domestic_worker              8.790323
business_owner               8.271688
manual_worker                7.276465

Now, let's look at how the conversion rate varies across different occupations (showing just a few for initial insights):

* student: 28.67% - This is the highest conversion rate among the listed occupations,
* retired: 22.79% - Second highest conversion rate among the listed occupations,
* jobless: 15.50% - Third highest conversion rate among the listed occupations,
* Executive: 13.76%
* Administrative Staff: 12.20%
* Independent Worker: 11.84%, and other occupations, with a lower conversion rate.

In dependence of strategy or marketing goals, we can consider these groups to be more likely to convert.

## Education Level

In [36]:
conversion_rate_education = conversion_rate_by('education_level')
print("Conversion rates by education level:")
conversion_rate_education.head(10).sort_values(by='conversion_rate', ascending=False)

Conversion rates by education level:


conversion_status  conversion_rate
education_level                   
college                  15.008647
unidentified             13.570275
high_school              10.560800
elementary_school         8.627737

Individuals with a college education have the highest conversion rate among the educational groups, which might indicate a better receptiveness or understanding of the offers presented in the campaign. In the future wold be interesting to determine educational level for "unidentified" individuals for a more accurate analysis.

## Marital Status

In [37]:
conversion_rate_marital = conversion_rate_by('marital_status')
print("\nConversion rates by marital status (%):")
conversion_rate_marital.head().sort_values(by='conversion_rate', ascending=False)


Conversion rates by marital status (%):


conversion_status  conversion_rate
marital_status                    
single                   14.951517
divorced                 11.945458
married                  10.124954

Single individuals show a higher conversion rate compared to married and divorced individuals. This may likely be due to the fact that married people make decisions together when it comes to finances.

## Communication Channel

In [38]:
conversion_rate_channel = conversion_rate_by('communication_channel')
print("\nConversion rates by communication channel:")
conversion_rate_channel.head().sort_values(by='conversion_rate', ascending=False)


Conversion rates by communication channel:


conversion_status      conversion_rate
communication_channel                 
mobile                       14.920429
landline                     13.420509
unidentified                  4.071599

Mobile communications seem to be the most effective, with the highest conversion rate, suggesting a potential focus area for future campaigns.

## Call Month

In [39]:
conversion_rate_month = conversion_rate_by('call_month')
print("\nConversion rates by call month:")
conversion_rate_month.head(12).sort_values(by='conversion_rate', ascending=False)


Conversion rates by call month:


conversion_status  conversion_rate
call_month                        
March                    51.991614
December                 46.728972
September                46.459413
October                  43.766938
April                    19.679400
February                 16.647792
August                   11.016813
June                     10.226634
November                 10.151134
January                  10.121169
July                      9.094865
May                       6.719942

The highest conversion rates occur in March, December, September and October, indicating that timing might play a crucial role in the effectiveness of the campaign.

## Age groups

In [40]:
# Display conversion rates for other attributes
conversion_rate_age_bins = conversion_rate_by(pd.cut(data['age'], bins=[0, 30, 40, 50, 60, 100], labels=['<30', '30-40', '40-50', '50-60', '60+']))
print("Conversion rates by age group:")
conversion_rate_age_bins.head()


Conversion rates by age group:


conversion_status  conversion_rate
age                               
<30                      16.289657
30-40                    10.247710
40-50                     9.066643
50-60                    10.053304
60+                      42.255892

The age group 60+ has a significantly higher conversion rate compared to other age groups, suggesting that older individuals are much more likely to convert than younger ones.

# Call Effectiveness
### Average Call Duration for Conversions vs. Non-Conversions

In [41]:
# Calculate the average call duration for conversions vs. non-conversions
average_duration = data.groupby('conversion_status')['call_duration'].mean()
print("Average call duration for conversions vs. non-conversions:")
average_duration


Average call duration for conversions vs. non-conversions:


conversion_status
converted        537.294574
not_converted    221.199870
Name: call_duration, dtype: float64

__Converted:__ 537.29 seconds (approximately 9 minutes)

__Not Converted:__ 221.20 seconds (approximately 3.7 minutes)

### Optimal Call Frequency:

In [42]:
# Group by call frequency and calculate the conversion rate at each frequency directly
conversion_rates_call_frequency = data.groupby('call_frequency')['conversion_status'].agg(
    conversion_rates_call_frequency=lambda x: (x == 'converted').mean()
).reset_index()

# Determine the call frequency with the highest conversion rate
optimal_frequency = conversion_rates_call_frequency.loc[conversion_rates_call_frequency['conversion_rates_call_frequency'].idxmax()]
print(f"Optimal call frequency: {optimal_frequency}\n\n")

conversion_rates_call_frequency.head(5).sort_values(by='conversion_rates_call_frequency', ascending=False)

Optimal call frequency: call_frequency                     1.000000
conversion_rates_call_frequency    0.145992
Name: 0, dtype: float64




call_frequency  conversion_rates_call_frequency
0               1                         0.145992
1               2                         0.112053
2               3                         0.111936
3               4                         0.090057
4               5                         0.078798

The analysis shows that the optimal number of calls for the highest conversion rate is 1 call with a conversion rate of approximately 14.6%. Conversion rates decrease as the number of calls increases, which suggests that additional calls beyond the first do not significantly increase the likelihood of conversion. 

Here's a quick look at the trend:

- 1 call: 14.6% conversion rate
- 2 calls: 11.2% conversion rate
- 3 calls: 11.2% conversion rate
- 4 calls: 9.0% conversion rate
- 5 calls: 7.9% conversion rate
- And further decreases as calls increase

This data suggests that most successful conversions occur with fewer calls, and making more than one call typically results in diminishing returns. This pattern reinforces the strategy that focusing on the quality of the first call might be more beneficial than increasing the number of calls.

### Conversion rate analysis by Month and Day of Month

In [43]:
# Conversion Rate by Month
conversion_by_month = data[data['conversion_status'] == 'converted'].groupby('call_month').size() / data.groupby('call_month').size()
conversion_by_month = conversion_by_month.reset_index(name='conversion_rate').sort_values(by='conversion_rate', ascending=False).head(12)
conversion_by_month.head(12)

call_month  conversion_rate
7       March         0.519916
2    December         0.467290
11  September         0.464594
10    October         0.437669
0       April         0.196794
3    February         0.166478
1      August         0.110168
6        June         0.102266
9    November         0.101511
4     January         0.101212
5        July         0.090949
8         May         0.067199

In [44]:
# Conversion Rate by Day of the Month
conversion_by_day = data[data['conversion_status'] == 'converted'].groupby('call_day').size() / data.groupby('call_day').size()
conversion_by_day = conversion_by_day.reset_index(name='conversion_rate').sort_values(by='conversion_rate', ascending=False).head(31)
conversion_by_day.head(31)

call_day  conversion_rate
0          1         0.279503
9         10         0.230916
29        30         0.173052
21        22         0.170166
2          3         0.164968
3          4         0.159170
24        25         0.158333
11        12         0.152215
12        13         0.152050
1          2         0.140867
14        15         0.139753
23        24         0.138702
15        16         0.135689
22        23         0.134185
26        27         0.133809
10        11         0.122380
8          9         0.114744
13        14         0.113636
4          5         0.112565
25        26         0.112077
7          8         0.109239
20        21         0.099210
17        18         0.098787
5          6         0.093685
16        17         0.090768
6          7         0.086406
27        28         0.078185
28        29         0.073926
30        31         0.071540
19        20         0.069767
18        19         0.069476

#### By Month

The months with the highest conversion rates are __March__, __December__, and __September__ with conversion rates of __March (52.0%)__, __December (46.7%)__, and __September (46.5%)__. This suggests that campaigns held in these months might be particularly effective.
Conversely, the months with the lowest conversion rates are __June__ and __July__ with conversion rates of __May (6.7%)__, __July (9.1%)__, and __January (10.1%)__. Campaigns during these months might need different strategies or expectations.

#### By Day

The days at the beginning of the month, particularly the __1st day (27.95%)__ and the __10th day (23.09%)__, show higher conversion rates. There is also a notable spike on the __30th (17.31%)__.
In general, the conversion rates are higher at the start of the month and on specific days later in the month (e.g., the __22nd__ and __30th__), indicating possible optimal days for reaching out.

#### Seasonal Trends:

The high conversion rates in March, September, and December suggest a seasonal influence, potentially tied to specific economic or social factors like end-of-quarter financial decisions or holiday seasons. The low rates in the summer months (especially May and July) could be due to vacations and lower availability or interest in business discussions.
These insights could guide the strategic timing of marketing campaigns, suggesting a focus on spring and late fall for optimal effectiveness, while being cautious or adjusting strategies during the summer and early winter.

#### Conversion Rate Analysis by Quarter
Let's analyze the conversion rates by quarters to observe more consolidated seasonal trends. We'll categorize the months into quarters and then calculate the conversion rates for each quarter. This should provide a clearer picture of the best times in the year for campaigns.

In [45]:
# Mapping from months to quarters
month_to_quarter = {
    'January': 'Q1', 'February': 'Q1', 'March': 'Q1',
    'April': 'Q2', 'May': 'Q2', 'June': 'Q2',
    'July': 'Q3', 'August': 'Q3', 'September': 'Q3',
    'October': 'Q4', 'November': 'Q4', 'December': 'Q4'
}

# Assign quarters based on the month
data['quarter'] = data['call_month'].map(month_to_quarter)

# Conversion Rate by Quarter - Optimized
conversion_by_quarter = data.groupby('quarter')['conversion_status'].agg(lambda x: (x == 'converted').mean()).reset_index(name='conversion_rate')

# Retrieve the top 4 entries (since there are only 4 quarters)
conversion_by_quarter.head(4)


quarter  conversion_rate
0      Q1         0.183484
1      Q2         0.092939
2      Q3         0.115469
3      Q4         0.167818

__Q1__ shows the highest conversion rates, which is consistent with our earlier findings showing strong performance in March.
__Q4__ also performs well, particularly influenced by high conversion rates in October and December.
__Q2__ and __Q3__ have lower conversion rates, with __Q3__ being the strongest and __Q2__   being the weakest.
These insights suggest that focusing marketing efforts during __Q1__ and __Q4__ could yield better results, while campaigns in __Q2__ and __Q3__
 might require different strategies or increased efforts to achieve similar outcomes.

### The influence of previous campaign outcomes on the current campaign's conversion rates.

In [46]:
# Conversion Rate by Previous Campaign Outcome
conversion_by_prev_outcome_corrected = (
    data.groupby('previous_campaign_outcome')['conversion_status']
    .agg(lambda x: (x == 'converted').mean())
    .reset_index(name='conversion_rate')
)

conversion_by_prev_outcome_corrected


previous_campaign_outcome  conversion_rate
0             other_outcome         0.166848
1                successful         0.647253
2              unidentified         0.091630
3              unsuccessful         0.126097

The analysis of the influence of previous campaign outcomes on the current campaign's success is as follows:
- __Successful:__ When the previous campaign was successful, the current campaign shows a high conversion rate of about 64.7%. This suggests a strong positive impact of prior success on current outcomes.
- __Unsuccessful:__ Previous campaigns that were unsuccessful result in a conversion rate of 12.6% for the current campaign, which is higher than might be expected but still significantly lower than following a successful campaign.
- __Unidentified:__ If the outcome of the previous campaign was unidentified, the conversion rate is 9.16%, indicating a neutral or unclear previous impact.
- __Other Outcome:__ For those categorized under "other outcome", the conversion rate is 16.7%, which suggests a moderate positive effect compared to unidentified outcomes but significantly lower than successful previous outcomes.

These results highlight the importance of achieving success in campaigns as it significantly boosts the potential for future conversions. Moreover, even campaigns identified as unsuccessful or other outcomes perform better than those with unidentified results, which could point to the value of clear outcomes and learning from past efforts.

In [47]:
# Categorize the 'age' column into specified groups
data['age_group'] = pd.cut(data['age'], 
                           bins=[0, 29, 40, 50, 60, 120], 
                           labels=['<30', '30-40', '40-50', '50-60', '60+'], 
                           right=True)

# Display the distribution of the new 'age_group' column to verify categorization
age_group_distribution = data['age_group'].value_counts()
age_group_distribution


age_group
30-40    19439
40-50    11239
50-60     8067
<30       5272
60+       1188
Name: count, dtype: int64

In [48]:
from scipy.stats import chi2_contingency

# Create contingency table for 'age_group' and 'conversion_status'
age_conversion_table = pd.crosstab(data['age_group'], data['conversion_status'])

# Conduct chi-squared test
chi2_age, p_age, dof_age, expected_age = chi2_contingency(age_conversion_table)

print(f"Chi-squared test results: chi2 = {chi2_age:.2f}, p-value = {p_age:.4f}")

Chi-squared test results: chi2 = 1378.01, p-value = 0.0000


In [49]:
# Create contingency tables for 'education_level' and 'marital_status'
education_conversion_table = pd.crosstab(data['education_level'], data['conversion_status'])
marital_conversion_table = pd.crosstab(data['marital_status'], data['conversion_status'])

# Conduct chi-squared tests
chi2_education, p_education, dof_education, expected_education = chi2_contingency(education_conversion_table)
chi2_marital, p_marital, dof_marital, expected_marital = chi2_contingency(marital_conversion_table)
print(f"Chi-squared test results:\nchi2_education = {chi2_education:.2f}, p-value_education = {p_education:.4f}\nchi2_marital = {chi2_marital:.2f}, p-value_marital = {p_marital:.4f}") 


Chi-squared test results:
chi2_education = 238.93, p-value_education = 0.0000
chi2_marital = 196.51, p-value_marital = 0.0000


In [50]:
marital_conversion_table.head()

conversion_status  converted  not_converted
marital_status                             
divorced                 622           4585
married                 2755          24455
single                  1912          10876

In [51]:
education_conversion_table.head()

conversion_status  converted  not_converted
education_level                            
college                 1996          11303
elementary_school        591           6259
high_school             2450          20749
unidentified             252           1605

In [52]:
# Create contingency table for 'occupation' and 'conversion_status'
occupation_conversion_table = pd.crosstab(data['occupation'], data['conversion_status'])

# Conduct chi-squared test
chi2_occupation, p_occupation, dof_occupation, expected_occupation = chi2_contingency(occupation_conversion_table)

print(f"Chi-squared test results: chi2 = {chi2_occupation:.2f}, p-value = {p_occupation:.4f}")

Chi-squared test results: chi2 = 835.84, p-value = 0.0000


In [53]:
occupation_conversion_table.head(10)

conversion_status     converted  not_converted
occupation                                    
administrative_staff        631           4539
business_owner              123           1364
domestic_worker             109           1131
executive                  1301           8156
independent_worker          187           1392
jobless                     202           1101
manual_worker               708           9022
retired_worker              516           1748
service_worker              369           3784
student                     269            669

## As next step we will start with Predictive Modeling
- Predictive Analytics: Build a predictive model to forecast the likelihood of a customer converting based on their profile and interaction history. This can help in prioritizing resources towards more likely prospects. 

To build a predictive model to forecast the likelihood of a customer converting based on their profile and interaction history, we can follow these steps:

1. Data Preparation
Feature Engineering: Convert categorical variables into numerical using encoding techniques like one-hot encoding.
Split Data: Divide the dataset into training and testing sets to ensure the model is tested on unseen data.
2. Model Selection
Choose Models: Start with baseline models such as logistic regression for binary classification and potentially explore more complex models like random forests or gradient boosting machines depending on the performance.
3. Model Training
Train the Model: Fit the model on the training data.
Parameter Tuning: Use techniques like grid search with cross-validation to find the best parameters for the model.
4. Model Evaluation
Evaluate the Model: Use metrics like accuracy, AUC-ROC, precision, recall, and F1-score to evaluate the performance of the model.
Validation: Validate the model on the testing set to check how well it generalizes to new data.
5. Deployment and Monitoring
Deploy the Model: Once validated, the model can be deployed for real-time predictions or batch predictions on new data.
Monitor and Update: Continuously monitor the model’s performance and update it as necessary when performance degrades or when new data becomes available.
Shall we proceed with these steps to develop the predictive model? If yes, we'll start with the data preparation phase, particularly focusing on feature engineering and splitting the data.

We'll perform the following tasks:
- __Encode Categorical Variables:__ We need to transform categorical variables into a format that can be used by machine learning models. This typically involves converting categories into numerical values using one-hot encoding or label encoding.

- __Split the Data:__ We will divide the dataset into a training set (usually about 70-80% of the data) and a testing set (20-30% of the data). This will allow us to train the model on a portion of the data and then test it on a separate set to evaluate its performance.

First, I'll take care of encoding the categorical variables. We'll look at the unique values in each categorical column to decide on the encoding method.
Let's start by listing the categorical columns and their unique values.

We have several categorical variables that we need to encode:

- __Occupation:__ 12 unique categories
- __Education Level:__ 4 unique categories
- __Marital Status:__ 3 unique categories
- __Communication Channel:__ 3 unique categories
- __Call Month:__ 12 unique categories
- __Previous Campaign Outcome:__ 4 unique categories
- __Conversion Status (target variable):__ 2 categories (This will be encoded as 0 or 1 for binary classification.)

For the categorical features, I'll use one-hot encoding because it avoids the implicit ordinal relationships that can be misinterpreted by the model (which could happen with label encoding).
For the target variable 'conversion_status', I'll use a simple label encoder to convert it into binary format (0 or 1).

In [54]:
# List of categorical columns and their unique values
categorical_cols = data.select_dtypes(include=['object']).columns
categorical_values = {col: data[col].unique() for col in categorical_cols}

The dataset has been successfully encoded, converting all categorical variables into numerical format suitable for model training. We now have 48 columns, which include one-hot encoded features and the binary target variable 'conversion_status'.

The next step is to split the dataset into training and testing sets. This is critical for training the model on one subset of the data and then validating its performance on a separate unseen subset. Let's proceed with splitting the data.

In [55]:
from sklearn.preprocessing import LabelEncoder

# Encode target variable using Label Encoder
label_encoder = LabelEncoder()
data['conversion_status'] = label_encoder.fit_transform(data['conversion_status'])

# One-hot encode categorical variables (excluding target variable 'conversion_status')
data_encoded = pd.get_dummies(data.drop('conversion_status', axis=1))

# Add the encoded target variable back to the dataframe
data_encoded['conversion_status'] = data['conversion_status']

# Display the first few rows of the encoded dataset to verify
data_encoded.head()


age  call_day  call_duration  call_frequency  \
0   28         9              1               1   
1   58         5            307               2   
2   40         4            113               1   
3   63         7             72               5   
4   43        29            184               4   

   occupation_administrative_staff  occupation_business_owner  \
0                             True                      False   
1                             True                      False   
2                            False                      False   
3                            False                      False   
4                            False                       True   

   occupation_domestic_worker  occupation_executive  \
0                       False                 False   
1                       False                 False   
2                       False                 False   
3                       False                 False   
4                       False                 False   

   occupation_independent_worker  occupation_jobless  ...  quarter_Q1  \
0                          False               False  ...       False   
1                          False               False  ...       False   
2                          False                True  ...        True   
3                          False               False  ...       False   
4                          False               False  ...       False   

   quarter_Q2  quarter_Q3  quarter_Q4  age_group_<30  age_group_30-40  \
0       False        True       False           True            False   
1        True       False       False          False            False   
2       False       False       False          False             True   
3        True       False       False          False            False   
4       False        True       False          False            False   

   age_group_40-50  age_group_50-60  age_group_60+  conversion_status  
0            False            False          False                  1  
1            False             True          False                  1  
2            False            False          False                  1  
3            False            False           True                  1  
4             True            False          False                  1  

[5 rows x 52 columns]

The data has been split into training and testing sets. We have:

* Training set: 31,643 samples
* Testing set: 13,562 samples

    Each set contains 42 features.

Now that the data is prepared, we can proceed to the model selection phase. For a baseline, we can start with a simple logistic regression model, which is widely used for binary classification problems. After establishing a baseline, we can explore more complex models if necessary.

In [56]:
from sklearn.model_selection import train_test_split

# Separate features and target variable
X = data_encoded.drop('conversion_status', axis=1)
y = data_encoded['conversion_status']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Check the shape of the training and testing data
X_train.shape, X_test.shape, y_train.shape, y_test.shape


((31643, 51), (13562, 51), (31643,), (13562,))

In [57]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, roc_auc_score

# Initialize and train the logistic regression model
model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)

# Predict on the training data
y_train_pred = model.predict(X_train)
y_train_proba = model.predict_proba(X_train)[:, 1]

# Predict on the testing data
y_test_pred = model.predict(X_test)
y_test_proba = model.predict_proba(X_test)[:, 1]

# Evaluate the model
train_accuracy = accuracy_score(y_train, y_train_pred)
test_accuracy = accuracy_score(y_test, y_test_pred)
train_auc = roc_auc_score(y_train, y_train_proba)
test_auc = roc_auc_score(y_test, y_test_proba)

train_accuracy, test_accuracy, train_auc, test_auc


(0.901937237303669, 0.9015631912697243, 0.9082974093115542, 0.9026228335571646)

The logistic regression model has performed as follows:

* Training Accuracy: 90.19%
* Testing Accuracy: 90.16%
* Training AUC-ROC: 90.82%
* Testing AUC-ROC: 90.26%

These results suggest that the model is fairly consistent and not overfitting, as the performance on both the training and testing sets is similar.

The AUC-ROC scores are also high, indicating good discriminative ability between the classes.

#### We will proceed eit trying another model.

### Random Forests Classifier

In [58]:
from sklearn.ensemble import RandomForestClassifier

# Initialize and train the Random Forest model
rf_model = RandomForestClassifier(n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train)

# Predict on the training and testing data
y_train_rf_pred = rf_model.predict(X_train)
y_train_rf_proba = rf_model.predict_proba(X_train)[:, 1]
y_test_rf_pred = rf_model.predict(X_test)
y_test_rf_proba = rf_model.predict_proba(X_test)[:, 1]

# Evaluate the Random Forest model
train_rf_accuracy = accuracy_score(y_train, y_train_rf_pred)
test_rf_accuracy = accuracy_score(y_test, y_test_rf_pred)
train_rf_auc = roc_auc_score(y_train, y_train_rf_proba)
test_rf_auc = roc_auc_score(y_test, y_test_rf_proba)

train_rf_accuracy, test_rf_accuracy, train_rf_auc, test_rf_auc


(1.0, 0.9047338150715234, 1.0, 0.9225225901789611)

The Random Forest model has provided the following performance metrics:

- Training Accuracy: 100%
- Testing Accuracy: 90.57%
- Training AUC-ROC: 100%
- Testing AUC-ROC: 92.10%

    The model achieves perfect accuracy and AUC on the training set, indicating it has learned the training data thoroughly, possibly to the point of overfitting. However, the performance on the test set is still robust, especially with an AUC-ROC of over 92%, which is an improvement over the logistic regression model.

Given the potential overfitting observed (perfect scores on training data), we might consider:

__Tuning the hyperparameters of the Random Forest__, such as reducing the number of trees, adjusting the max depth, or increasing the minimum samples split to generalize better.

__Feature selection__ to reduce the number of features and potentially improve model generalization.

We will proceed with with hyperparameter tuning

In [59]:
from sklearn.model_selection import GridSearchCV

# Define the parameter grid
param_grid = {
    'n_estimators': [50, 100, 200],
    'max_depth': [10, 20, None]
}

# Create the grid search model
rf_grid = GridSearchCV(estimator=RandomForestClassifier(random_state=42),
                       param_grid=param_grid,
                       cv=3,  # 3-fold cross-validation
                       scoring='roc_auc',  # Evaluate using AUC-ROC
                       verbose=2)

# Fit the grid search to the data
rf_grid.fit(X_train, y_train)

# Best parameters and best score
best_params = rf_grid.best_params_
best_score = rf_grid.best_score_
best_params, best_score


Fitting 3 folds for each of 9 candidates, totalling 27 fits
[CV] END ......................max_depth=10, n_estimators=50; total time=   0.7s
[CV] END ......................max_depth=10, n_estimators=50; total time=   0.6s
[CV] END ......................max_depth=10, n_estimators=50; total time=   0.9s
[CV] END .....................max_depth=10, n_estimators=100; total time=   1.7s
[CV] END .....................max_depth=10, n_estimators=100; total time=   1.6s
[CV] END .....................max_depth=10, n_estimators=100; total time=   1.5s
[CV] END .....................max_depth=10, n_estimators=200; total time=   3.2s
[CV] END .....................max_depth=10, n_estimators=200; total time=   2.8s
[CV] END .....................max_depth=10, n_estimators=200; total time=   2.7s
[CV] END ......................max_depth=20, n_estimators=50; total time=   1.0s
[CV] END ......................max_depth=20, n_estimators=50; total time=   1.2s
[CV] END ......................max_depth=20, n_es

({'max_depth': 20, 'n_estimators': 200}, 0.9277876107239891)

Ways to Implementing the Model
- Model Setup: Configure the Random Forest classifier using these optimal parameters.
- Model Training: Train the model on your full training dataset to maximize the learning scope.
- Evaluation: Validate the model again on your test dataset to confirm the performance improvements (like checking AUC-ROC, accuracy, precision, recall, and F1-score).
- Interpretation: Analyze the model results, understanding which features are most influential in predicting outcomes.
 
This can be done by examining feature importance's provided by the Random Forest algorithm.

In [67]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, roc_auc_score, confusion_matrix, classification_report

# Initialize the Random Forest model with the optimal parameters
model = RandomForestClassifier(n_estimators=200, max_depth=20, random_state=42)

# Train the model on the training data
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test)
y_proba = model.predict_proba(X_test)[:, 1]

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
roc_auc = roc_auc_score(y_test, y_proba)
conf_mat = confusion_matrix(y_test, y_pred)
class_report = classification_report(y_test, y_pred)

print("Accuracy:", accuracy)
print("ROC AUC Score:", roc_auc)
print("Confusion Matrix:\n", conf_mat)
print("Classification Report:\n", class_report)


Accuracy: 0.9067246718773042
ROC AUC Score: 0.928449444629528
Confusion Matrix:
 [[  628   937]
 [  328 11669]]
Classification Report:
               precision    recall  f1-score   support

           0       0.66      0.40      0.50      1565
           1       0.93      0.97      0.95     11997

    accuracy                           0.91     13562
   macro avg       0.79      0.69      0.72     13562
weighted avg       0.89      0.91      0.90     13562



### Summary of Results:
#### Accuracy: 90.67%

The model correctly predicted the outcome for 90.67% of the cases in the test dataset, which is a solid overall performance.

#### ROC AUC Score: 92.84%

The Area Under the Receiver Operating Characteristic Curve (AUC-ROC) is 92.84%, indicating a high ability of the model to distinguish between the two classes (converted vs. not converted). This is an excellent score, suggesting that the model has a strong predictive capability.

#### Confusion Matrix:

- True Negative (TN) = 628: The number of correctly predicted non-conversions.
- False Positive (FP) = 937: The number of non-conversions incorrectly predicted as conversions.
- False Negative (FN) = 328: The number of conversions incorrectly predicted as non-conversions.
- True Positive (TP) = 11,669: The number of correctly predicted conversions.

    This matrix shows that while the model is very good at identifying true positives (conversions), it struggles relatively more with correctly classifying the negatives (non-conversions).

Precision, Recall, and F1-Score:

* Class 0 (Non-conversions):
- Precision: 66% - Of all predicted non-conversions, 66% were actual non-conversions.
- Recall: 40% - Of all actual non-conversions, only 40% were correctly identified.
- F1-Score: 50% - The harmonic mean of precision and recall for non-conversions.

* Class 1 (Conversions):
- Precision: 93% - Of all predicted conversions, 93% were actual conversions.
- Recall: 97% - Of all actual conversions, 97% were correctly identified.
- F1-Score: 95% - The harmonic mean of precision and recall for conversions.

#### Overall Scores:
- Macro Average F1-Score: 72% - Average F1-score unweighted by class size.
- Weighted Average F1-Score: 90% - Average F1-score weighted by the number of true instances for each class.

#### Conclusions:

The model is very effective in predicting conversions, evidenced by high precision and recall for the majority class (conversions).

It has less effectiveness in identifying non-conversions, as seen by lower recall and precision for the minority class.

The high AUC-ROC score indicates strong separability between classes under varying thresholds.

The overall accuracy and weighted F1-score suggest that the model performs very well on the majority class but could benefit from improvements in identifying non-conversions more accurately.

Using this Random Forest model, you can classify a list of clients by priority based on their likelihood of conversion.

This is particularly useful for marketing strategies where resources need to be allocated efficiently to target customers who are more likely to convert.

#### Here’s how you can achieve this:

__Step 1:__ Predict Conversion Probabilities

First, the model can predict the probability of conversion for each client in your list. This involves feeding the clients' profile and interaction history data into the model, which then outputs a probability score for each client.

__Step 2:__ Rank Clients

You can then rank the clients based on these probabilities. Clients with higher scores are more likely to convert and should be prioritized in marketing campaigns. This ranking allows you to focus your efforts and resources on the clients with the highest potential return.

__Step 3:__ Strategic Engagement

Based on the ranking, you can develop tiered marketing strategies:

- **High Priority:** Clients with the highest probability of conversion might receive more personalized and direct marketing efforts.
- **Medium Priority:** Those with moderate scores could be nurtured with automated follow-ups and regular engagement.
- **Lower Priority:** Clients with low probabilities might receive less frequent or general promotional materials, which helps in managing resources without completely ignoring any segment.

#### Example Implementation

For example, if you have a new campaign ready to launch, you could:

__Prepare the Input Data:__ Gather and preprocess data for all clients as done during the model training phase, including encoding categorical variables and normalizing where necessary.

__Predict Using the Model:__ Use the Random Forest model to predict conversion probabilities.

__Sort the Output:__ Sort clients by their predicted probabilities in descending order.

__Execute Campaigns:__ Tailor the campaign intensity and resources according to the sorted list, targeting the top-ranked clients first.

This approach maximizes efficiency by focusing resources on individuals who are most likely to respond positively, enhancing the overall effectiveness of marketing efforts.


## Model Serialization

In [68]:
import pickle

# Save the model to disk
filename = 'finalized_model.sav'
pickle.dump(model, open(filename, 'wb'))

# Load the model from disk later for prediction
loaded_model = pickle.load(open(filename, 'rb'))


#### Set Up the Production Environment

**Cloud Deployment:** Server Setup: Decide whether the model will run on a cloud platform (AWS, Azure, Google Cloud) or on-premises. Set up the necessary servers, databases, and API gateways.

**API Development:** Develop an API using frameworks like Flask or FastAPI in Python. This API will handle requests from the front end, process them with the model, and return predictions.